File: hw/gudino/u3/hw35StrategyPersistence/StrategyPy/controller/SortContext.py
from controller.BubbleSort import BubbleSort
from controller.InsertionSort import InsertionSort
from controller.QuickSort import QuickSort

class SortContext:
    """
    AQUÍ SE IMPLEMENTA EL PATRÓN STRATEGY
    El Context decide qué algoritmo (Strategy) usar
    """

    def __init__(self, data, repository):
        self.data = data
        self.repository = repository
        self.strategy = self._choose_strategy()

    def _choose_strategy(self):
        size = len(self.data)
        if 2 <= size <= 5:
            return BubbleSort()
        elif 6 <= size <= 10:
            return InsertionSort()
        elif size >= 11:
            return QuickSort()
        else:
            raise ValueError("Array must have more than 1 element.")

    def sort(self):
        original = self.data.copy()   # ← COPIA del input original (NO se modifica)
        result = self.strategy.sort(self.data)

        self.repository.save(
            original,
            result,
            self.strategy.__class__.__name__
        )
        return result

File: exam/gudino/u3/StrategyJavaGudino/src/main/java/controller/SortContext.java
package controller;

import model.*;

/**
 * CONTEXT Strategy
 */
public class SortContext {

    private SortStrategy strategy;
    private SortRepository repository;

    public SortContext(int[] data, SortRepository repository) {
        this.repository = repository;
        this.strategy = chooseStrategy(data.length);
    }
 
    private SortStrategy chooseStrategy(int size) {
        if (size <= 6) {
            return new BubbleSort();
    } else if (size <= 10) {
        return new InsertionSort();
    } else {
        return new QuickSort();
    }
}

    public int[] sort(int[] data) {
        int[] original = data.clone();   
        int[] sorted = strategy.sort(data);

        repository.save(original, sorted, getStrategyName());
        return sorted;
    }

    public String getStrategyName() {
        return strategy.getClass().getSimpleName();
    }
}
